staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -452,6 +452,19 @@ func getNewItemFunc(listObj runtime.Object, v reflect.Value) func() runtime.Obje
 
 func (s *store) Count(key string) (int64, error) {
 	key = path.Join(s.pathPrefix, key)
+
+	// Consider the following keys:
+	// A: <storage-prefix>//foo.bar.io/machines
+	// B: <storage-prefix>//foo.bar.io/machinesets
+	// If we ask for the count of key A, the result will also include the count for key B
+	// since key B shares the same prefix as key A.
+	//
+	// Add a separator to mark the end of the key here, this will exclude all other
+	// keys from a different resource that is a prefix of the specified key.
+	if len(key) > 0 && key[len(key)-1] != '/' {
+		key = fmt.Sprintf("%s/", key)
+	}
```

Comment:
use the same code as List():

https://github.com/kubernetes/kubernetes/blob/f64f687fcfb22e5099fbcb990f3e6faa44446fff/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L548-L553